blog-posts/lambda-extensions/readme.md

```diff
@@ -0,0 +1,443 @@
+
+import { EventTypes, ExtensionAPIService } from 'lambda-extension-service';
+import { Log } from './types';
+import { LogAggregator } from './logAggregator';
+import { listenForLog } from './logServer';
+import { LambdaContext } from './lambdaContext';
+import { forwardLogs } from './forwardLogs';
+
+console.log('Executing monitor extension code...');
+
+const main = async () => {
+  const logAggregator = new LogAggregator();
+  const lambdaContext = new LambdaContext();
+  const onLogReceived = (log: Log) => {
+    logAggregator.addLog(log, lambdaContext.getRequestId());
+  };
+  listenForLog(onLogReceived);
+
+  const extensionApiService = new ExtensionAPIService({
+    extensionName: 'monitor',
+  });
+  await extensionApiService.register([EventTypes.Invoke, EventTypes.Shutdown]);
+
+  while (true) {
+    const event = await extensionApiService.next();
+    const lastContext = lambdaContext.getContext();
+    lambdaContext.updateContext(event);
+
+    if (lastContext !== undefined) {
+      await forwardLogs({
+        context: lastContext,
+        logs: logAggregator.getLogs(lastContext.requestId),
+      });
+    }
+  }
+};
+
+main().catch(error => console.error(error));
+```
+
+#### 🚀 Deploy the external extension
+
+First, we also need to make the extension code executable by the lambda. But this time it should be a standalone executable file.
+
+💡 We reuse the same `esbuild` command as with the interceptor extension, but we need to add a shebang to make the file executable with the `--banner:js='#!/usr/bin/env node'` option.
+
+```json
+{
+  "scripts": {
+    "build:monitor": "./node_modules/.bin/esbuild  src/layers/monitorExtension/monitor/index.ts --bundle --outfile='./dist/layers/monitorExtension/monitor.js' --platform=node --main-fields=module,main --banner:js='#!/usr/bin/env node'"
+  }
+}
+```
+
+Then we need to ship the code to the lambda thanks to a lambda layer.
+
+💡 By outputting the build of the monitor code in the same folder as the interceptor code (`./dist/layers/interceptorExtension`), the layer will ship both files without changing our CDK configuration.
+
+Finally, we need to make the lambda execute the monitor process.
+
+By default, the lambda service will try to execute all files in the `/opt/extensions` folder when starting the lambda environment.
+
+So let's add a bash script into `./dist/layers/monitorExtension/extensions` to start the monitor code.
+
+```bash
+# dist/layers/monitorExtension/extensions/monitor
+
+#!/bin/bash
+set -euo pipefail
+
+OWN_FILENAME="$(basename $0)"
+LAMBDA_EXTENSION_NAME="$OWN_FILENAME" # (external) extension name has to match the filename
+NODE_OPTIONS="" # Needed to reset NODE_OPTIONS set by Lambda runtime. Otherwise, the internal interceptor extension will be loaded in the external process too.
+
+exec "/opt/${LAMBDA_EXTENSION_NAME}.js"
+```
+
+This script will be shipped to `/opt/extensions/monitor` and will be executed automatically by the lambda service.
```

Comment:
This script will be delivered to /opt/extensions/monitor and will automatically be executed by the Lambda service.

💡Take note that the environment variables of the Lambda are shared across processes. Consequently, the external process will have NODE_OPTIONS set to require the interceptor code. It's crucial to reset NODE_OPTIONS to avoid loading the interceptor code twice, which would inadvertently intercept the monitor calls.

With that, we've successfully created our monitoring extension.

👀 Results
Now, let's redeploy it on our previous Lambda. Here are the final logs:

monitored lambda execution logs

🟠 The Lambda service initializes the Lambda execution environment.
🟢 The external monitoring extension kicks into action.
🟣 The internal interceptor extension comes to life.
🟠 The Lambda service confirms the successful initialization of the monitoring extension.
🟠 The Lambda service commences the first event handling.
🟣 The internal interceptor extension intercepts the two HTTP calls.
🟠 The Lambda service concludes the first event handling.
🟠 The Lambda service launches the second event handling.
🟢 The external monitoring extension dispatches the logs of the first event to the external monitoring service.
🟣 The internal interceptor extension intercepts the two HTTP calls again.
🟠 The Lambda service wraps up the first event handling.
🟢 Upon Lambda shutdown, the external monitoring extension dispatches the logs of the second event to the external monitoring service.
On the faux external monitoring service, we've received event-specific, aggregated logs 🥳:

forwarded logs

⏭️ Trial it on your own AWS account by using the dedicated repository:

{% embed https://github.com/CorentinDoue/lambda-extensions-example %}

Conclusion
You now understand how to create both internal and external extensions for Lambda. This method provides a robust way to build and distribute support tools for your Lambdas without influencing their performance or altering the handler code crafted by other developers.

I encourage you to share your insights and inquiries in the comments section below. I'll be more than happy to respond.